Kubelet CPU manager: Consider option alignBySocket while DistrubuteCPUsAcrossNUMA

ruquanzhao opened the pull request:
Signed-off-by: Ruquan Zhao ruquan.zhao@arm.com

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:

/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/114642

#### Special notes for your reviewer: 
Now `takeByTopologyNUMADistributed` prefers the NUMANodes combination in same socket if `alighBySocket` is set to true.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Commit:
Consider option alignBySocket while DistrubuteCPUsAcrossNUMA

Signed-off-by: Ruquan Zhao ruquan.zhao@arm.com

bart0sh commented:
/triage accepted
/priority important-longterm
/assign @swatisehgal @ffromani @klueska 
/ok-to-test

ffromani commented:
/release-note fix interaction between align-by-socket and distribute-cpus-across-numa cpumanager policy options

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -245,6 +245,11 @@ func (d CPUDetails) CPUsInCores(ids ...int) cpuset.CPUSet {
+// IsNUMANodesInSameSocket returns true for all NUMANodes in the same socket
+func (d CPUDetails) IsNUMANodesInSameSocket(NUMANodes []int) bool {
+	return d.SocketsInNUMANodes(NUMANodes...).Size() <= 1
+}
```
let's please add unit test coverage for this helper

ffromani commented on pkg/kubelet/cm/cpumanager/policy_test.go:
```diff
@@ -100,6 +100,31 @@ var (
 		},
 	}
 
+	topoDualSocketDualNUMAPerSocketNoHT = &topology.CPUTopology{
```
it would be nice to add a comment to record if this topology represents an existing CPU or if it has been made up to exercise the code. If possible, let's prefer topologies of existing silicon.

ffromani replied:
scratch this, you did in the followup commit. We're good.

ffromani reviewed (commented):
thanks for this PR, looks promising. I need to re-read the original issue, meantime added a couple preliminary comments.

Commit:
Add existing silicon and unittest.

Signed-off-by: Ruquan Zhao ruquan.zhao@arm.com

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -715,10 +716,22 @@ func takeByTopologyNUMADistributed(topo *topology.CPUTopology, availableCPUs cpu
 			// combo (and remainder set) to be the best one found so far.
-			if bestLocalBalance < bestBalance {
+			if alignBySocket {
+				if topo.CPUDetails.IsNUMANodesInSameSocket(combo) && !bestBalanceInOneSocket {
```
nit: I'd rename as `AreNUMANodesInSameSocket`

ruquanzhao replied:
Thanks! Done.

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -1147,7 +1147,7 @@ func TestPolicyWithAlignBySocketAndDistributeCPUsAcrossNUMAEnabled(t *testing.T)
 		{
 			description: "AlignBySocket and DistributeCPUsAcrossNUMA both enabled",
-			topo:        topoDualSocketDualNUMAPerSocketNoHT,
+			topo:        topoDualSocketMultiNumaPerSocketHTType2,
```
perhaps rename `topoDualSocketMultiNumaPerSocketSublusterHT`? or in general something more descriptive than `Type2`

ffromani commented on pkg/kubelet/cm/cpumanager/policy_test.go:
```diff
@@ -540,6 +515,108 @@ var (
 		},
 	}
+	/*
+		Topology type 2 from dual xeon gold 6230; lscpu excerpt
```
"topology with numa sub clustering"?

ruquanzhao replied:
Base the information I have searched(maybe I am wrong), both this new Topology and Topology of Gold 6230 now
 in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_test.go#L418 have SNC enabled. 
If it's hard to name this, I think we could load similar silicon to 6230, such as Gold 6226(https://www.spec.org/cpu2017/results/res2019q3/cpu2017-20190722-16283.html).

What do you think? @ffromani

ffromani replied:
looks good enough, no need to change then

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -715,10 +716,22 @@ func takeByTopologyNUMADistributed(topo *topology.CPUTopology, availableCPUs cpu
 			// lowest "balance score" of all previous combos, then update this
 			// combo (and remainder set) to be the best one found so far.
-			if bestLocalBalance < bestBalance {
+			if alignBySocket {
```
would it be possible to reformat the code to avoid nested ifs and use `return Continue` like the original code is? If possible it should be a bit more readable/consistent

ruquanzhao replied:
Done.

ffromani reviewed (commented):
thanks @ruquanzhao for your work! the happy path LGTM. Let's make sure we have enough coverage for the other paths (see inline comments). Last but not least, I'd make one (last?) attempt to simplify the nested ifs. But the latter is minor, the main and only remaining concern is to make sure the new code works as expected on the other paths. We may need some more test coverage. Other than that, LGTM

swatisehgal reviewed (commented):
The PR looks good to me but I see some review comments are yet to be addressed.

@ruquanzhao Can you address review comments from @ffromani that are still pending to be addressed (nested ifs and rename `IsNUMANodesInSameSocket` to `AreNUMANodesInSameSocket`) ?

klueska commented:
I'll wait to review this in detail until I get lgtms from @ffromani and @swatisehgal

Commit:
Add testcases when both AlighBySocket and
DistributeCPUsAcrossNUMAEnabled.
Change the calculation of "distribution" when allocate CPUs across NUMA.

Signed-off-by: Ruquan Zhao ruquan.zhao@arm.com

ffromani commented:
/release-note none

ffromani commented:
> /release-note none

It seems I don't have the permissions to change this. @ruquanzhao could you please fix the release note? I'll review this PR again in the coming days

ruquanzhao commented:
/release-note none

ruquanzhao commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -715,10 +716,22 @@ func takeByTopologyNUMADistributed(topo *topology.CPUTopology, availableCPUs cpu
+					bestRemainder = bestLocalRemainder
+					bestCombo = combo
+					bestBalanceInOneSocket = true
+				} else if topo.CPUDetails.IsNUMANodesInSameSocket(combo) == bestBalanceInOneSocket && bestLocalBalance < bestBalance {
```
Done.

ffromani commented:
release note still unfixed (this will prevent merge even with LGTM+approve)
plus the unit test failure looks legit

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -591,7 +591,7 @@ func takeByTopologyNUMADistributed(topo *topology.CPUTopology, availableCPUs cpu
 			// If we've already found a combo with a balance of 0 in a
 			// different iteration, then don't bother checking any others.
-			if bestBalance == 0 {
+			if bestBalance == 0 && bestBalanceInOneSocket {
```
this is set to true only if alignBySocket is true. Do we have a bug? or are we "just" doing more iterations than before?

Commit:
Fix unittest.

Signed-off-by: Ruquan Zhao ruquan.zhao@arm.com

ruquanzhao commented:
/release-note-none

ruquanzhao commented:
/retest

bart0sh commented:
@ruquanzhao please address pending review comments: https://github.com/kubernetes/kubernetes/pull/118068#pullrequestreview-1512585632 to proceed further with this PR, thanks.

bart0sh commented:
/cc

ruquanzhao commented:
> The PR looks good to me but I see some review comments are yet to be addressed.
> 
> @ruquanzhao Can you address review comments from @ffromani that are still pending to be addressed (nested ifs and rename `IsNUMANodesInSameSocket` to `AreNUMANodesInSameSocket`) ?

Thanks for the review! Done.

ruquanzhao commented:
@bart0sh Thanks for the kindly remind. I have apply the changes in previous changes.

swatisehgal reviewed (commented):
@ruquanzhao Thanks for your work on this and addressing all the review comments.

The PR looks good. It would be really nice if you can add some node e2e tests where `alignBySocket` and `DistrubuteCPUsAcrossNUMA` policy options are enabled independently and when both are enabled for us to ensure expected behavior and prevent regressions in the future. You can refer to [CPU Manager node e2e tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go). We currently only have e2e tests for `FullPCPUsOnlyOption` and adding tests for these policy options would bring us one step closer to graduating these policy options to Beta.

ffromani commented:
I think we're getting there. I'd have a final pass later. @ruquanzhao was kind enough to also fix the e2e cpumanager serial lane in https://github.com/kubernetes/test-infra/pull/30545 . Let's make sure the e2e test pass and let's try to add few more. The latter could be challenging due to the high HW requirements, but still worth a shot.

ruquanzhao commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ruquanzhao commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2
